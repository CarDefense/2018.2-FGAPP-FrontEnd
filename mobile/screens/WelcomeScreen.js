import React, { Component } from "react";
import { View, StyleSheet, Image, FlatList, Text, TouchableOpacity, ImageBackground } from 'react-native';
import { TextField } from 'react-native-material-textfield';
import MaterialIcon from 'react-native-vector-icons/MaterialIcons';
import jwt_decode from 'jwt-decode';
import { PROFILE_API } from './tab_navigator/car_defense/screens/TabNavigator/const/Const.js';

class WelcomeScreen extends Component {

    async logIn() {
        const { type, token } = await Expo.Facebook.logInWithReadPermissionsAsync('2177002552568068', {
            permissions: ['public_profile'],
        });
        if (type === 'success') {
            // Get the user's name using Facebook's Graph API
            const response = await fetch(`https://graph.facebook.com/v3.2/me?access_token=${token}&debug=all&fields=id%2Cname%2Cemail&format=json&method=get&pretty=0&suppress_http_code=1`);
            const user = (await response.json());
            console.log(user);
            this.props.navigation.navigate('TabHandler', { user: user });
        }
    }

    static navigationOption = {
        header: 'none'
    }
    constructor(props) {
        super(props);

        this.onFocus = this.onFocus.bind(this);
        this.onAccessoryPress = this.onAccessoryPress.bind(this);

        this.usernameRef = this.updateRef.bind(this, 'username');
        this.passwordRef = this.updateRef.bind(this, 'password');

        this.renderPasswordAccessory = this.renderPasswordAccessory.bind(this);

        this.state = {
            userName: '',
            password: '',
            secureTextEntry: true,
            non_field_alert: ['']
        };

    }

    onFocus() {
        let { errors = {} } = this.state;

        for (let name in errors) {
            let ref = this[name];

            if (ref && ref.isFocused()) {
                delete errors[name];
            }
        }

        this.setState({ errors });
    }

    onChangeText(text) {
        ['username', 'password']
            .map((name) => ({ name, ref: this[name] }))
            .forEach(({ name, ref }) => {
                if (ref.isFocused()) {
                    this.setState({ [name]: text });
                }
            });
    }

    onAccessoryPress() {
        this.setState(({ secureTextEntry }) => ({ secureTextEntry: !secureTextEntry }));
    }

    updateRef(name, ref) {
        this[name] = ref;
    }

    renderPasswordAccessory() {
        let { secureTextEntry } = this.state;

        let name = secureTextEntry ?
            'visibility' :
            'visibility-off';

        return (
            <MaterialIcon
                size={24}
                name={name}
                color={TextField.defaultProps.baseColor}
                onPress={this.onAccessoryPress}
                suppressHighlighting
            />
        );
    }

    _onPressButton = async () => {
        let errors = {};
        let errorUserName = false;
        let errorPassword = false;

        ['username']
            .forEach((text) => {
                let value = this[text].value();

                if (!value) {
                    errors[text] = 'Digite o nome de usuário!';
                    errorUserName = true;
                }
            });

        ['password']
            .forEach((text) => {
                let value = this[text].value();

                if (!value) {
                    errors[text] = 'Digite a senha!';
                    errorPassword = true;
                }
                else {
                    if (value.length < 8) {
                        errors[text] = 'Senha muito curta.';
                    }
                    else if (value.length > 15) {
                        errors[text] = 'Senha muito longa.';
                    }
                }
            });
            
        if (!errorUserName && !errorPassword) {
            var login_path = PROFILE_API + '/token-auth/';
            fetch(login_path, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    'username': this.state.username,
                    'password': this.state.password,
                }),
            })
            .then((response) => response.json())
            .then((responseJson) => {
                console.log(JSON.stringify(responseJson));
                if (responseJson.non_field_errors != undefined){
                    this.setState({ non_field_alert: ['Usuário ou senha incorreto(s).']})
                }
                else{
                    this.setState({ non_field_alert: ['']})
                }
                var token = responseJson.token ? responseJson.token : undefined;
                user = token ? jwt_decode(token) : undefined;
                if (user != undefined || responseJson.key != undefined) {
                    this.props.navigation.navigate('TabHandler', { user: user });
                }
            })
            .catch(err => {
                if (typeof err.text === 'function') {
                    err.text().then(errorMessage => {
                        this.props.dispatch(displayTheError(errorMessage))
                    });
                } else {
                    Alert.alert("Erro na conexão.");
                }
            });
        }
        this.setState({ errors });

    }

    render() {
        let { errors = {}, secureTextEntry, ...data } = this.state;

        return (
            <ImageBackground
                source={{uri: 'http://cardefense2.eastus.cloudapp.azure.com:8002/media/b9_VpUTIV2.png'}}
                style={{ width: '100%', height: '100%' }}
            >
                <View style={styles.container}>
                    <View style={styles.containerImage}>
                        <Image
                            style={styles.image}
                            source={{uri: 'http://cardefense2.eastus.cloudapp.azure.com:8002/media/icontest_T8bTMAG.png'}}
                        />
                    </View>
                    <View style={styles.container1}>
                        <TextField
                            ref={this.usernameRef}
                            value={data.username}
                            autoCorrect={false}
                            enablesReturnKeyAutomatically={true}
                            onFocus={this.onFocus}
                            onChangeText={(username) => this.setState({ username })}
                            returnKeyType='next'
                            label='Nome de usuário'
                            tintColor="white"
                            underlineColorAndroid="transparent"
                            error={errors.username}
                            textColor='white'
                        />
                        <TextField
                            ref={this.passwordRef}
                            value={data.password}
                            secureTextEntry={secureTextEntry}
                            autoCapitalize='none'
                            autoCorrect={false}
                            enablesReturnKeyAutomatically={true}
                            clearTextOnFocus={true}
                            onFocus={this.onFocus}
                            onChangeText={(password) => this.setState({ password })}
                            returnKeyType='done'
                            label='Senha'
                            tintColor="white"
                            underlineColorAndroid="transparent"
                            error={errors.password}
                            maxLength={20}
                            characterRestriction={15}
                            renderAccessory={this.renderPasswordAccessory}
                            textColor='white'
                        />
                        <View style={styles.containerButton}>
                            <TouchableOpacity
                                style={styles.button}
                                onPress={() => this._onPressButton()}
                                containerViewStyle={{ width: '40%' }}
                            >
                                <Text style={{ color: 'white', fontSize: 15, fontWeight: '300' }} >LOG IN</Text>
                            </TouchableOpacity>
                            <FlatList
                                data={this.state.non_field_alert}
                                renderItem={({item}) => <Text style ={{color: 'red'}}>{item}</Text>}
                                keyExtractor={item => 'non_field_errors'}
                            />
                        </View>
                    
                        <TouchableOpacity
                            style={styles.buttonFacebook}
                            onPress={() => this.logIn()}
                            containerViewStyle={{ width: '40%' }}
                        >
                            <Text style={{ color: 'white', fontWeight: '700'}} >Facebook</Text>
                        </TouchableOpacity>
                        <TouchableOpacity
                            style={styles.button1}
                            onPress={() => this.props.navigation.navigate('SignUpScreen')}
                            containerViewStyle={{ width: '40%' }}
                        >
                            <Text style={{ color: "#8bd4da" }} >Criar conta</Text>
                        </TouchableOpacity>
                    </View>
                </View>
            </ImageBackground>
        );
    }
}
export default WelcomeScreen;

const styles = StyleSheet.create({
    button: {
        backgroundColor: "#8bd4da",
        borderRadius: 15,
        height: 40,
        width: 320,
        justifyContent: 'center',
        alignItems: 'center'
    },
    container1: {
        marginTop: 150,
      },
    buttonFacebook: {
        backgroundColor: "#3B5998",
        borderRadius: 15,
        height: 40,
        width: 100,
        justifyContent: 'center',
        alignItems: 'center'
    },
    button1: {
        backgroundColor: "rgba(0, 0, 0, 0)",
        borderRadius: 15,
        height: 40,
        width: 120,
        alignItems: 'center'
    },
    container: {
        margin: 16,
    },
    containerButton: {
        alignItems: 'center',
        marginTop: 10
    },
    containerImage: {
        margin: 30,
        paddingTop: 4,
        paddingRight: 4,
        alignItems: 'center'
    },
    containerText: {
        margin: 60,
        alignItems: 'center'
    },
    image: {
        height: 100,
        width: 100,
    }
});