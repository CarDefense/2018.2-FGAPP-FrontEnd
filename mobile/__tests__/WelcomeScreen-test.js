import WelcomeScreen from '../screens/WelcomeScreen'
import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

test('WelcomeScreen snapShot', () => {
    const snap = renderer.create(
        <WelcomeScreen/>
    ).toJSON()
    expect(snap).toMatchSnapshot()
})