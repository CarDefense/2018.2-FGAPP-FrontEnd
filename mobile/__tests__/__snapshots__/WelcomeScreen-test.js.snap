// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WelcomeScreen snapShot 1`] = `
<View
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "http://cardefense2.eastus.cloudapp.azure.com:8002/media/b9_VpUTIV2.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "margin": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "margin": 30,
          "paddingRight": 4,
          "paddingTop": 4,
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "http://cardefense2.eastus.cloudapp.azure.com:8002/media/icontest_T8bTMAG.png",
          }
        }
        style={
          Object {
            "height": 100,
            "width": 100,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "marginTop": 150,
        }
      }
    >
      <View
        onResponderRelease={[Function]}
        onStartShouldSetResponder={[Function]}
        pointerEvents="auto"
        style={undefined}
      >
        <View
          collapsable={undefined}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "rgba(0, 0, 0, 0.38)",
              "borderBottomWidth": 0.5,
              "height": 64,
              "paddingBottom": 8,
              "paddingTop": 32,
            }
          }
        >
          <View
            collapsable={undefined}
            style={
              Object {
                "position": "absolute",
                "top": 36,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              collapsable={undefined}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.38)",
                  "fontSize": 16,
                }
              }
            >
              Nome de usu√°rio
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="sentences"
              autoCorrect={false}
              disableFullscreenUI={true}
              editable={true}
              enablesReturnKeyAutomatically={true}
              error={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onContentSizeChange={[Function]}
              onFocus={[Function]}
              returnKeyType="next"
              selectionColor="white"
              style={
                Array [
                  Object {
                    "flex": 1,
                    "margin": 0,
                    "padding": 0,
                    "top": 2,
                  },
                  Object {
                    "color": "white",
                    "fontSize": 16,
                    "height": 24,
                    "textAlign": "left",
                  },
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "flexDirection": "row",
              "height": 8,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgb(213, 0, 0)",
                    "fontSize": 0,
                    "opacity": 0,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(0, 0, 0, .38)",
                    "fontSize": 12,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onResponderRelease={[Function]}
        onStartShouldSetResponder={[Function]}
        pointerEvents="auto"
        style={undefined}
      >
        <View
          collapsable={undefined}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "rgba(0, 0, 0, 0.38)",
              "borderBottomWidth": 0.5,
              "height": 64,
              "paddingBottom": 8,
              "paddingTop": 32,
            }
          }
        >
          <View
            collapsable={undefined}
            style={
              Object {
                "position": "absolute",
                "top": 36,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              collapsable={undefined}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.38)",
                  "fontSize": 16,
                }
              }
            >
              Senha
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="none"
              autoCorrect={false}
              disableFullscreenUI={true}
              editable={true}
              enablesReturnKeyAutomatically={true}
              error={undefined}
              maxLength={20}
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onContentSizeChange={[Function]}
              onFocus={[Function]}
              renderAccessory={[Function]}
              returnKeyType="done"
              secureTextEntry={true}
              selectionColor="white"
              style={
                Array [
                  Object {
                    "flex": 1,
                    "margin": 0,
                    "padding": 0,
                    "top": 2,
                  },
                  Object {
                    "color": "white",
                    "fontSize": 16,
                    "height": 24,
                    "textAlign": "left",
                  },
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
            <View
              style={
                Object {
                  "alignSelf": "flex-start",
                  "justifyContent": "center",
                  "top": 2,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
              />
            </View>
          </View>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "flexDirection": "row",
              "height": 24,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgb(213, 0, 0)",
                    "fontSize": 0,
                    "opacity": 0,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(0, 0, 0, .38)",
                    "fontSize": 12,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
          <View
            style={
              Object {
                "paddingLeft": 4,
                "paddingVertical": 4,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "textAlign": "right",
                  },
                  undefined,
                  Object {
                    "color": "rgba(0, 0, 0, .38)",
                    "fontSize": 12,
                  },
                ]
              }
            >
              0
               / 
              15
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "marginTop": 10,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#8bd4da",
              "borderRadius": 15,
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "width": 320,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
                "fontSize": 15,
                "fontWeight": "300",
              }
            }
          >
            LOG IN
          </Text>
        </View>
        <RCTScrollView
          data={
            Array [
              "",
            ]
          }
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          invertStickyHeaders={undefined}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View>
            <View
              onLayout={[Function]}
              style={null}
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "red",
                  }
                }
              >
                
              </Text>
            </View>
          </View>
        </RCTScrollView>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#3B5998",
            "borderRadius": 15,
            "height": 40,
            "justifyContent": "center",
            "opacity": 1,
            "width": 100,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "white",
              "fontWeight": "700",
            }
          }
        >
          Facebook
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0)",
            "borderRadius": 15,
            "height": 40,
            "opacity": 1,
            "width": 120,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#8bd4da",
            }
          }
        >
          Criar conta
        </Text>
      </View>
    </View>
  </View>
</View>
`;
